core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -59,24 +61,101 @@ object RequestChannel extends Logging {
+      // convert requestMetricsSizeBucketsConfig from string to int array. e.g., "0,1,10,50,100" => [0,1,10,50,100]
+      var bucketsString = requestMetricsSizeBucketsConfig.replaceAll("\\s", "").split(',')
+      if(bucketsString.length < 2) {
+        error(s"requestMetricsSizeBuckets has fewer than 2 buckets: '${requestMetricsSizeBucketsConfig}'. " +
+          s"use default value instead ")
+        bucketsString = Defaults.RequestMetricsSizeBuckets.replaceAll("\\s", "").split(',')
+      }
+      var bucketsInt = Array[Int](bucketsString.length)
+      try {
+        bucketsInt = bucketsString.map(_.toInt)
+      } catch {
+        case e: Exception => {
+          error(s"Failed to parse requestMetricsSizeBuckets: '${requestMetricsSizeBucketsConfig}'. " +
+            s"use default value instead ", e)
+          bucketsInt = Defaults.RequestMetricsSizeBuckets.replaceAll("\\s", "").split(',').map(_.toInt)
+        }
+      }
+
+      // get size and corresponding term to be used in metric name
+      // [0,1,10,50,100] => requestSizeBuckets:Seq((0, "0To1Mb"), (1, "1To10Mb"), (10, "10To50Mb"), (50, "50To100Mb"),
+      // (100, "100MbGreater"))
+      val requestSizeBuckets = for (i <- 0 until bucketsInt.length) yield {
+        val size = bucketsInt(i)
+        if (i == bucketsInt.length - 1) (size, s"${size}MbGreater")
+        else (size, s"${size}To${bucketsInt(i + 1)}Mb")
+      }
+      requestSizeBuckets.map(bucket => (bucket._1, s"${requestType}${bucket._2}")).toMap
+    }
+
+    // generate map[acks, map[requestSize, metricName]] for produce requests of different request size and acks
+    def getProduceRequestAcksSizeMetricNameMap():Map[Int, Map[Int, String]] = {
+      val produceRequestAcks = Seq((0, "0"), (1, "1"), (-1, "All"))
+      val requestSizeMetricNameMap = getRequestSizeMetricNameMap(ApiKeys.PRODUCE.name)
+
+      produceRequestAcks.map(ack =>
+        (ack._1, requestSizeMetricNameMap.map({case(size, name) => (size, s"${name}Acks${ack._2}")}))).toMap
+    }
+
+    // generate map[responseSize, metricName] for consumerFetch requests of different request size and acks
+    def getConsumerFetchRequestAcksSizeMetricNameMap():Map[Int, String] = {
+      getRequestSizeMetricNameMap(RequestMetrics.consumerFetchMetricName)
+    }
+
+    // get all the metric names for produce requests of different acks and size
+    def getProduceRequestAcksSizeMetricNames : Seq[String] = {
+      produceRequestAcksSizeMetricNameMap.values.toSeq.map(a => a.values.toSeq).flatten
+    }
+
+    // get all the metric names for fetch requests of different size
+    def getConsumerFetchRequestSizeMetricNames : Seq[String] = {
+      consumerFetchRequestSizeMetricNameMap.values.toSeq
+    }
+
+    // get the metric name for a given request/response size
+    // the bucket is [left, right)
+    def getRequestSizeBucketMetricName(sizeMetricNameMap: Map[Int, String], sizeBytes: Long): String = {
```

Comment:
It seems using a NavigatableMap and the corresponding floorEntry() API could make the logic more readable.

Reply:
Changed to use Java TreeMap to make this part of logic simple, but it added some conversion between scala and java collection. I think scala 2.13 would have methods maxBefore and minAfter which are more suitable to be used here.